[Fix-17355] Fix reassignWorkflowInstanceHost may failed when no events in channel

reele opened the pull request:
…nceId existence.

<!--Thanks very much for contributing to Apache DolphinScheduler, we are happy that you want to help us improve DolphinScheduler! -->

close #17355

<!--(For example: This pull request adds checkstyle plugin).-->

## Brief change log

Modify reassignWorkflowInstanceHost's return value based on taskInstanceId existence instead of based on channel existence.

## Verify this pull request

<!--*(Please pick either of the following options)*-->

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

<!--*(example:)*
- *Added dolphinscheduler-dao tests for end-to-end.*
- *Added CronUtilsTest to verify the change.*
- *Manually verified the change by testing locally.* -->

(or)

## Pull Request Notice
[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

If your pull request contains incompatible change, you should also add it to `docs/docs/en/guide/upgrade/incompatible.md`

Commit:
Modify reassignWorkflowInstanceHost's return value based on taskInstanceId existence.

Commit:
Modify reassignWorkflowInstanceHost's return value based on taskInstanceId existence.

Commit:
Modify reassignWorkflowInstanceHost's return value based on taskInstanceId existence.

Commit:
Modify reassignWorkflowInstanceHost's return value based on taskInstanceId existence.

reele committed:
Merge branch 'dev' into fix-reassign-host-issue

ruanwenjun commented on dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/executor/PhysicalTaskEngineDelegator.java:
```diff
@@ -76,9 +78,14 @@ public boolean reassignWorkflowInstanceHost(final TaskExecutorReassignMasterRequ
-        physicalTaskExecutorRepository.get(taskInstanceId).ifPresent(
-                taskExecutor -> taskExecutor.getTaskExecutionContext().setWorkflowInstanceHost(workflowHost));
-        return physicalTaskExecutorEventReporter.reassignWorkflowInstanceHost(taskInstanceId, workflowHost);
+        final Optional<ITaskExecutor> taskExecutorOptional = physicalTaskExecutorRepository.get(taskInstanceId);
+        physicalTaskExecutorEventReporter.reassignWorkflowInstanceHost(taskInstanceId, workflowHost);
+        if (taskExecutorOptional.isPresent()) {
+            final ITaskExecutor taskExecutor = taskExecutorOptional.get();
+            taskExecutor.getTaskExecutionContext().setWorkflowInstanceHost(workflowHost);
+            return true;
+        }
+        return false;
```
```suggestion
        Optional<ITaskExecutor> iTaskExecutorOptional = physicalTaskExecutorRepository.get(taskInstanceId);
        if (iTaskExecutorOptional.isPresent()) {
            iTaskExecutorOptional.get().getTaskExecutionContext().setWorkflowInstanceHost(workflowHost);
            return true;
        }
        return false;
```

It's better to only change the workflow instance host in `TaskExecutionContext`, and inject `ITaskExecutorRepository` into `TaskExecutorLifecycleEventRemoteReporter`. 

We can remove workflowInstanceHost from IReportableTaskExecutorLifecycleEvent, as this event does not inherently require knowledge of the workflow instance host. Instead, when reporting the event, the reporter should retrieve the current host from a centralized repository. This approach simplifies workflow host management and helps prevent potential concurrency issues. Once the workflow host changed, the reporter can get the correct host when report a event. 

Or we can set the workflow instance host into `ReportableTaskExecutorLifecycleEventChannel`, then we need to register the channel into reporter when the TaskExecutor registered, and change the host in channel when `reassignWorkflowInstanceHost`

reele replied:
ok.

reele replied:
get host address from TaskExecutor may not work, TaskExecutor's lifecycle and event's lifecycle are not completely overlapping, TaskExecutor will removed in onTaskExecutorFinalizeLifecycleEvent before the all events be reported.

but the channel‘s lifecycle is similar to event, it still needs to lock.

ruanwenjun replied:
`TaskExecutorFinalizeLifecycleEvent` is the end event, and it doesn't rely on RPC, it's only a local event(It's used to do some resource clean work), so it's ok to assembly the host to channel?

reele replied:
yes, i mean, i did test for get host address from task-executor, when reporter handle the failed or other final events, the task-executor is already removed from repository.

assembly the host to channel is not ok, channel will removed when current events are all reported, at this moment reassign host will not work, this is a lifecycle overlapping issue,  Channel or task-executor must have one to cover the task's all lifecycle to hold the host.

i'm thinking to keep the channel's lifecycle when it's empty, and inject ITaskExecutorRepository into TaskExecutorLifecycleEventRemoteReporter use to check the task-executor exists, if channel is empty and task-executor was removed, then remove the channel.

ruanwenjun replied:
Your idea is OK for me, from my perspective, the channel's lifecycle should equal with task executor, they should register together and deregister together.

On the other hand, there shouldn't have any event after `TaskExecutorFinalizeLifecycleEvent`, so this seems a bug if the failed or other final event hasn't been handled, but the task executor is removed from repository, since only `TaskExecutorFinalizeLifecycleEvent` fired then will clean the task executor.

reele replied:
I think it because reporter and coordinator are in different thread, they are async to handle the event, so the task executor might be removed prematurely.

ruanwenjun replied:
A TaskExecutor will only be removed after the TaskExecutorFinalizeLifecycleEvent is fired, and this event should be fired in synchronous mode.

reele replied:
> A TaskExecutor will only be removed after the TaskExecutorFinalizeLifecycleEvent is fired, and this event should be fired in synchronous mode.

so, how about moving the publishing of TaskExecutorFinalizeLifecycleEvent to TaskExecutorLifecycleEventRemoteReporter? When a task reaches a terminal state (Success/Killed/Paused/Failed), we could publish a new TaskExecutorFinishedLifecycleEvent to mark the TaskExecutor as invalid or inactive. Then, the final event TaskExecutorFinalizeLifecycleEvent would instead be published by TaskExecutorLifecycleEventRemoteReporter.

I've submitted the new changes implemented according to this flow.

ruanwenjun replied:
Good idea. After the TaskExecutor received the ACK of a terminal event, then means the task executor's lifecycle is finished, then publish a Finalize event to clean the resource.

reele replied:
submitted.

Commit:
report event use TaskExecutor's host, expand the lifecycle of TaskExecutor to contain the channel's lifecycle.

Commit:
report event use TaskExecutor's host, expand the lifecycle of TaskExecutor to contain the channel's lifecycle.

Commit:
Merge remote-tracking branch 'origin/fix-reassign-host-issue' into fix-reassign-host-issue

reele committed:
Merge branch 'apache:dev' into fix-reassign-host-issue

Commit:
add TaskExecutorFinishedLifecycleEvent,
TaskExecutorFinalizeLifecycleEvent will be published by TaskExecutorLifecycleEventRemoteReporter.

Commit:
add TaskExecutorFinishedLifecycleEvent,
TaskExecutorFinalizeLifecycleEvent will be published by TaskExecutorLifecycleEventRemoteReporter.

Commit:
Merge remote-tracking branch 'origin/fix-reassign-host-issue' into fix-reassign-host-issue

Commit:
add TaskExecutorFinishedLifecycleEvent,
TaskExecutorFinalizeLifecycleEvent will be published by TaskExecutorLifecycleEventRemoteReporter.

Commit:
add TaskExecutorFinishedLifecycleEvent,
TaskExecutorFinalizeLifecycleEvent will be published by TaskExecutorLifecycleEventRemoteReporter.

reele commented on dolphinscheduler-extract/dolphinscheduler-extract-master/src/main/java/org/apache/dolphinscheduler/extract/master/TaskExecutorEventRemoteReporterClient.java:
```diff
@@ -33,32 +33,38 @@
 public class TaskExecutorEventRemoteReporterClient implements ITaskExecutorEventRemoteReporterClient {
 
-    public void reportTaskExecutionEventToMaster(final IReportableTaskExecutorLifecycleEvent taskExecutorLifecycleEvent) {
+    public void reportTaskExecutionEventToMaster(final String masterAddress,
```
fixed.

ruanwenjun commented on dolphinscheduler-task-executor/src/main/java/org/apache/dolphinscheduler/task/executor/events/TaskExecutorFinishedLifecycleEvent.java:
```diff
@@ -0,0 +1,42 @@
+@EqualsAndHashCode(callSuper = true)
+@SuperBuilder
+@NoArgsConstructor
+public class TaskExecutorFinishedLifecycleEvent extends AbstractTaskExecutorLifecycleEvent {
```
Please don't import a new Event type here. We don't really need this.

reele replied:
resubmitted.

ruanwenjun commented on dolphinscheduler-task-executor/src/main/java/org/apache/dolphinscheduler/task/executor/eventbus/TaskExecutorLifecycleEventRemoteReporter.java:
```diff
@@ -164,6 +169,24 @@ public Map<Integer, ReportableTaskExecutorLifecycleEventChannel> getEventChannel
         return eventChannels;
     }
 
+    /**
+     * Extend the lifecycle of the TaskExecutor to span the entire processing cycle of the task.
+     * before TaskExecutor changes to inactive, all events will put into channel,
+     * so we can just remove the TaskExecutor after the associated channel has been removed.
+     */
+    private void finalizeFinishedTaskExecutors() {
+        eventChannelsLock.lock();
+        // check if the taskExecutor is inactive and no channel exist, then we can publish finalize event.
+        for (final ITaskExecutor taskExecutor : taskExecutorRepository.getAllWaitingReport()) {
+            final Integer taskExecutorId = taskExecutor.getId();
+            if (!eventChannels.containsKey(taskExecutorId)) {
+                taskExecutorRepository.finishReport(taskExecutorId);
+                taskExecutor.getTaskExecutorEventBus().publish(TaskExecutorFinalizeLifecycleEvent.of(taskExecutor));
+            }
+        }
+        eventChannelsLock.unlock();
+    }
+
```
This is too complex. It's better to publish `TaskExecutorFinalizeLifecycleEvent` in `receiveTaskExecutorLifecycleEventACK` after receive a finished ack.

reele replied:
wow... it's pretty much better!

reele replied:
resubmitted.

ruanwenjun commented on dolphinscheduler-task-executor/src/main/java/org/apache/dolphinscheduler/task/executor/eventbus/TaskExecutorLifecycleEventRemoteReporter.java:
```diff
@@ -138,15 +148,10 @@ public void receiveTaskExecutorLifecycleEventACK(final TaskExecutorLifecycleEven
 
     @Override
-    public boolean reassignWorkflowInstanceHost(int taskInstanceId, String workflowHost) {
+    public void wake() {
```
It's better to revert this method, and change the implementation. Once the workflow host changed, we may need to reset the event's lastReportTime or set a new flag to mark the event need to retry now, and notify the reporter by `taskExecutionEventEmptyCondition.signalAll()`.

reele replied:
resubmitted.

Commit:
Merge remote-tracking branch 'origin/fix-reassign-host-issue' into fix-reassign-host-issue

Commit:
TaskExecutorFinalizeLifecycleEvent will be published by TaskExecutorLifecycleEventRemoteReporter.

Commit:
TaskExecutorFinalizeLifecycleEvent will be published by TaskExecutorLifecycleEventRemoteReporter.

Commit:
TaskExecutorFinalizeLifecycleEvent will be published by TaskExecutorLifecycleEventRemoteReporter.

ruanwenjun commented on dolphinscheduler-task-executor/src/main/java/org/apache/dolphinscheduler/task/executor/eventbus/TaskExecutorLifecycleEventRemoteReporter.java:
```diff
@@ -138,15 +152,14 @@ public void receiveTaskExecutorLifecycleEventACK(final TaskExecutorLifecycleEven
 
     @Override
-    public boolean reassignWorkflowInstanceHost(int taskInstanceId, String workflowHost) {
+    public void resetAndReadyChannelEvents(int taskInstanceId) {
```
```suggestion
    public void onWorkflowInstanceHostChanged(int taskInstanceId) {
```

reele replied:
apply committed

ruanwenjun reviewed (dismissed):
LGTM

Commit:
apply suggestion.

Commit:
fix api test issue

SbloodyS committed:
Merge branch 'dev' into fix-reassign-host-issue

SbloodyS reviewed (approved):
+1
